sdk/storage/Azure.Storage.Internal.Avro/src/AvroReader.cs

```diff
@@ -60,14 +60,28 @@ internal class AvroReader
         /// </summary>
         private bool _initalized;
 
+        /// <summary>
+        /// To detect redundant calls
+        /// </summary>
+        private bool _disposed = false;
+
         /// <summary>
         /// Constructor for an AvroReader that will read from the
         /// beginning of an Avro file.
         /// </summary>
         public AvroReader(Stream dataStream)
         {
-            _dataStream = dataStream;
-            _headerStream = dataStream;
+            if (dataStream.CanSeek)
+            {
+                _dataStream = dataStream;
+                _headerStream = dataStream;
+            }
+            else
+            {
+                _dataStream = new StreamWithPosition(dataStream);
```

Comment:
Can you track position as part of `AvroReader` itself?

Reply:
I could, but it would be extremely non-trivial.  Adding `StreamWithPosition` was much less disruptive, and seemed like the safer option.